### PR TITLE
ITL-2-Add-validation

### PR DESCRIPTION
**Pull request by Barabasbalazs**

### CODE REVIEW

File: src/middleware/validation.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  partialValidation: async (
+    req: Request,
+    res: Response,
+    next: NextFunction
+  ) => {
+    try {
+      await partialToDoValidator.validateAsync(req.body);
+      return next();
+    } catch (e) {
+      return next(errors.invalidParameter("Invalid ToDo"));
```
Comment (roli-wg) [inline src/middleware/validation.ts]: Please try to return more explanatory error messages, which is the first wrong field
Like "Missing or invalid parameter 'title'"

File: src/middleware/validation.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    }
+    if (sortBy && order) {
+      if (
+        isValueOfType(sortBy, sortByProperties) &&
+        isValueOfType(order, orderProperties)
+      ) {
+        return next();
+      }
+    }
+    return next(errors.invalidParameter("Invalid parameters"));
```
Comment (roli-wg) [inline src/middleware/validation.ts]: Please pay attention to error message here as well

File: src/routes/todos/todo-controller.ts [hunk 3]
```diff
@@ -21,9 +22,21 @@ export const todoController = {
...
+      { sort?: keyof ToDo; order?: OrderType; search: string }
+    >,
+    res: Response<ToDo[]>,
+    next: NextFunction
+  ) => {
     try {
-      const toDoList = await todoService.listAll();
+      const sortBy = req.query.sort;
+      const order = req.query.order;
+      const searchString = req.query.search;
```
Comment (roli-wg) [inline src/routes/todos/todo-controller.ts]: You can destructor the req.query object to define this variables in one line
`const { sortBy, order, searchString } = req.query`

File: src/services/todo-service.ts [hunk 4]
```diff
@@ -1,11 +1,27 @@
...
+      return await model.find({
+        $or: [
+          { title: { $regex: searchString } },
+          { text: { $regex: searchString } },
+        ],
+      });
+    }
+    if (sortBy && order) {
+      return await model.find().sort({ [sortBy]: order });
+    }
```
Comment (roli-wg) [inline src/services/todo-service.ts]: What if we want to search and sort in the same time, please rewrite this to support search and sort simultaneously

**Review by roli-wg — CHANGES_REQUESTED** [review]

**Review by roli-wg — APPROVED** [review]

File: src/routes/todos/todo-controller.ts [hunk 5]
```diff
@@ -21,21 +22,40 @@ export const todoController = {
     }
   },
 
-  getAll: async (req: Request, res: Response<ToDo[]>, next: NextFunction) => {
+  getAll: async (
+    req: Request<
+      {},
+      {},
+      {},
+      { sort?: keyof ToDo; order?: OrderType; search?: string }
```
Comment (komaromiGyorgy) [inline src/routes/todos/todo-controller.ts]: Can you make just create types for these? This looks gnarly wherever you have query requests, for example you could do: 

`
export interface QueryRequest<T extends { [queryParam: string]: string; }> extends Request {
    query: T
}
export interface BodyRequest<T> extends Request {
    body: T
}
export interface ParamRequest<T extends { [param: string]: string; [captureGroup: number]: string; }> extends Request {
    params: T
}

export function requesting(request: QueryRequest<ListRequest>) {
  request.query.order
}
`

**Review by komaromiGyorgy — APPROVED** [review]

File: src/index.ts [hunk 6]
```diff
@@ -20,7 +20,7 @@ app.use("/api", router);
 
 app.use(handleError);
 
-const port = process.env.PORT;
+const port = process.env.PORT ? process.env.PORT : 8080;
```
Comment (SzaszWebgurus) [inline src/index.ts]: const port = process.env.PORT || 8080

File: src/middleware/validation.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+import Joi from "joi";
+
+export const validators = {
+  toDoValidation: (dto: Joi.ObjectSchema<any>) => {
+    return async (req: Request, res: Response, next: NextFunction) => {
+      try {
+        await dto.validateAsync(req.body);
+        return next();
+      } catch (e) {
+        const errorReport = e as ErrorReport;
```
Comment (SzaszWebgurus) [inline src/middleware/validation.ts]: can you do this inline?

File: src/middleware/validation.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    };
+  },
+  queryValidation: async (
+    req: Request<{}, {}, {}, { sort?: any; order?: any }>,
+    res: Response,
+    next: NextFunction
+  ) => {
+    try {
+      const sortBy = req.query.sort;
+      const order = req.query.order;
```
Comment (SzaszWebgurus) [inline src/middleware/validation.ts]: const { sortBy, order} = req.query;

File: src/routes/todos/todo-controller.ts [hunk 9]
```diff
@@ -45,9 +70,17 @@ export const todoController = {
...
 
-  delete: async (req: Request, res: Response, next: NextFunction) => {
+  delete: async (
+    req: ParamRequest<{ id: string }>,
+    res: Response,
+    next: NextFunction
+  ) => {
     try {
-      const deleted = await todoService.remove(req.params._id);
+      const id = req.params.id;
```
Comment (SzaszWebgurus) [inline src/routes/todos/todo-controller.ts:76]: use general `validateId` DTO

File: src/routes/todos/todo-controller.ts [hunk 10]
```diff
@@ -60,13 +93,17 @@ export const todoController = {
...
 
   update: async (
-    req: Request<{ _id: string }, {}, Partial<ToDo>>,
+    req: ParamBodyRequest<Partial<ToDo>, { id: string }>,
     res: Response<ToDo>,
     next: NextFunction
   ) => {
     try {
+      const id = req.params.id;
+      if (!id) {
```
Comment (SzaszWebgurus) [inline src/routes/todos/todo-controller.ts]: same as above

File: src/routes/todos/todo-router.ts [hunk 11]
```diff
@@ -1,12 +1,22 @@
...
-todoRouter.patch("/:_id", todoController.update);
+todoRouter.post(
+  "/",
+  validators.toDoValidation(fullToDoDTO),
+  todoController.save
+);
+todoRouter.get("/", validators.queryValidation, todoController.getAll);
+todoRouter.get("/:id", todoController.getOne);
+todoRouter.delete("/:id", todoController.delete);
+todoRouter.patch(
```
Comment (SzaszWebgurus) [inline src/routes/todos/todo-router.ts]: put

File: src/routes/todos/todo-router.ts [hunk 12]
```diff
@@ -1,12 +1,22 @@
...
+  "/:id",
+  validators.toDoValidation(partialToDoDTO),
+  todoController.update
```
Comment (SzaszWebgurus) [inline src/routes/todos/todo-router.ts]: updateOne

File: src/services/todo-service.ts [hunk 13]
```diff
@@ -1,12 +1,33 @@
...
+    sortBy?: keyof ToDo,
+    order?: OrderType,
+    searchString?: string
+  ): Promise<ToDo[] | void> => {
+    const stages = [];
+    if (searchString) {
+      stages.unshift({
+        $match: {
+          $or: [
+            { title: { $regex: searchString } },
```
Comment (SzaszWebgurus) [inline src/services/todo-service.ts]: use the regex in case insensitive mode

File: src/services/todo-service.ts [hunk 14]
```diff
@@ -1,12 +1,33 @@
...
+            { text: { $regex: searchString } },
+          ],
+        },
+      });
+    } else {
+      stages.unshift({ $match: { title: { $regex: "/*" } } });
```
Comment (SzaszWebgurus) [inline src/services/todo-service.ts]: TODO: remove this <- add this comment

File: src/services/todo-service.ts [hunk 15]
```diff
@@ -1,12 +1,33 @@
...
+    }
+    if (sortBy && order) {
+      return await model.aggregate(stages).sort({ [sortBy]: order });
```
Comment (SzaszWebgurus) [inline src/services/todo-service.ts]: add the sort to the aggregation pipeline

**Review by SzaszWebgurus — CHANGES_REQUESTED** [review]

**Review by SzaszWebgurus — APPROVED** [review]

**Pull request merged**